### PR TITLE
[Refactor] lake::{Async}DeltaWriter creation using Builder pattern

### PR DESCRIPTION
**Pull request by sduzh**
Introduced lake::{Async}DeltaWriterBuilder class to unify the creation of {Async}DeltaWriter instances, instead of multiple overloaded constructors. This avoids confusion on which constructor to use.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by sduzh** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 71 / 114 (62.28%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/lake/delta_writer.h|15|38|39.47%|[115, 116, 120, 121, 122, 125, 126, 127, 135, 136, 137, 140, 141, 142, 150, 151, 152, 160, 161, 162, 175, 177, 178]|
|:large_blue_circle:|src/storage/lake/async_delta_writer.h|15|35|42.86%|[110, 112, 116, 117, 118, 121, 122, 123, 126, 127, 128, 136, 137, 138, 151, 152, 153, 166, 168, 169]|
|:large_blue_circle:|src/storage/lake/schema_change.cpp|7|7|100.00%|[]|
|:large_blue_circle:|src/runtime/lake_tablets_channel.cpp|13|13|100.00%|[]|
|:large_blue_circle:|src/storage/lake/delta_writer.cpp|4|4|100.00%|[]|
|:large_blue_circle:|src/storage/lake/async_delta_writer.cpp|17|17|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 71 / 114 (62.28%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/lake/delta_writer.h|15|38|39.47%|[115, 116, 120, 121, 122, 125, 126, 127, 135, 136, 137, 140, 141, 142, 150, 151, 152, 160, 161, 162, 175, 177, 178]|
|:large_blue_circle:|src/storage/lake/async_delta_writer.h|15|35|42.86%|[110, 112, 116, 117, 118, 121, 122, 123, 126, 127, 128, 136, 137, 138, 151, 152, 153, 166, 168, 169]|
|:large_blue_circle:|src/storage/lake/schema_change.cpp|7|7|100.00%|[]|
|:large_blue_circle:|src/runtime/lake_tablets_channel.cpp|13|13|100.00%|[]|
|:large_blue_circle:|src/storage/lake/delta_writer.cpp|4|4|100.00%|[]|
|:large_blue_circle:|src/storage/lake/async_delta_writer.cpp|17|17|100.00%|[]|

**Review by wyb — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

File: be/src/storage/lake/delta_writer.h [hunk 1]
```diff
@@ -128,7 +102,94 @@ class DeltaWriter {
...
 private:
+    DeltaWriter(TabletManager* tablet_manager, int64_t tablet_id, int64_t txn_id, int64_t partition_id,
+                const std::vector<SlotDescriptor*>* slots, const std::string& merge_condition,
+                bool miss_auto_increment_column, int64_t table_id, int64_t immutable_tablet_size,
+                MemTracker* mem_tracker, int64_t max_buffer_size);
+
     DeltaWriterImpl* _impl;
 };
 
+class DeltaWriterBuilder {
```
Comment (kevincai) [inline be/src/storage/lake/delta_writer.h:113]: This is flexible, but the downside is, given X parameters there could be 2^X combinations to construct the object, with which, there could be y combinations are valid. Others are all invalid. This builder doesn't look like take care of this situation, but assuming all the combinations are valid.

Comment (sduzh) [reply be/src/storage/lake/delta_writer.h:113]: Yes, I totally agree with you, I just haven't come up with a better solution, do you have any suggestions?

Comment (kevincai) [reply be/src/storage/lake/delta_writer.h:113]: maybe, make the target object constructor private, and move the check into Builder.build() with exception or nullptr.
or add some convenient interface to builder such as,
* Builder.setX().setY().buildForXXX(a, b, c)
* Builder.setX().setY().buildForYYY(d, e, f)

Comment (sduzh) [reply be/src/storage/lake/delta_writer.h:113]: Good suggestion, I can add some checks in Build.build().

**Pull request merged**